### PR TITLE
HUM-509 Ring management documentation

### PR DESCRIPTION
**Pull request by creiht**

### CODE REVIEW

File: docs/rings.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+The builder file contains extra information needed for building the rings and care should be taken to backup and not lose these files.
+
+## Adding Devices
+
+Each individual storage devices is stored in the ring.  Before adding the device, you will need to know the zone, server ip address and port, device name, and weight of the device.  The server ip address and port should be address and port that the server for that ring is running on.  The device name should be the name of the partitioned device.  The weight should be the number of Gigabytes the device can store.
+
+A device can be added with the command `hummingbird ring <builder_file> add add z<zone>-<ip>:<port>/<device_name>` and an example deivce added to the object ring might look like `hummingbird ring object.builder add z1-10.0.0.1:6000/xvdd 1000` for a 1TB device.
+
+Note:  As devices are added to the ring, the actual mappings do not change until you run the reblance command.  It is best to make all the changes you want to make before rebalancing.
```
Comment (dpgoetz) [inline docs/rings.md]: reblance

File: docs/rings.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+## Rebalancing the Ring
+
+The `hummingbird ring <builder_file> rebalance` command will take the information in the ring builder file and builde a compressed ring file that can be used by the servers.  When done, the comannd will return how many partitions moved and the balance of the new ring.
```
Comment (dpgoetz) [inline docs/rings.md]: builde

File: docs/rings.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
+# Ring Managment 101
+
+## Ring Overview
+
+The hummingbird rings are responsible for mapping what devices data is stored on.  There is a separate ring for the objects, containers and accounts. Rings are created when you build a new cluster and updated when changes to the cluster need to be made like for example, adding or removing nodes or devices to the storage cluster.  It is important that all servers in the cluster have the same and most current version of the ring.
```
Comment (corystone) [inline docs/rings.md]: Weird sentence. Maybe just "mapping data to devices" is clearer.

File: docs/rings.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+A ring can be generated with the command `hummingbird ring <builder_file> create <part_power> <replicas> <min_part_hours>` and an example object ring with the typical settings can be created with `hummingbird ring object.builder create 22 3 168`.
+
+The builder file contains extra information needed for building the rings and care should be taken to backup and not lose these files.
+
+## Adding Devices
+
+Each individual storage devices is stored in the ring.  Before adding the device, you will need to know the zone, server ip address and port, device name, and weight of the device.  The server ip address and port should be address and port that the server for that ring is running on.  The device name should be the name of the partitioned device.  The weight should be the number of Gigabytes the device can store.
+
+A device can be added with the command `hummingbird ring <builder_file> add add z<zone>-<ip>:<port>/<device_name>` and an example deivce added to the object ring might look like `hummingbird ring object.builder add z1-10.0.0.1:6000/xvdd 1000` for a 1TB device.
```
Comment (dpgoetz) [inline docs/rings.md]: deivce

File: docs/rings.md (hunk 3 above)
Comment (corystone) [reply docs/rings.md]: Oxford comma. (containers, and)

File: docs/rings.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+hummingbird ring account.builder add r1z2-10.1.1.11:6002/sdc1 500
+hummingbird ring account.builder add r1z3-10.1.1.12:6002/sdb1 500
+hummingbird ring account.builder add r1z3-10.1.1.12:6002/sdc1 500
+hummingbird ring account.builder rebalance
+```
+
+It is *very* important that you add all nodes and devices to the ring before running the rebalance command.
+
+## Handling a Device failure
+
```
Comment (dpgoetz) [inline docs/rings.md]: degradation

**Comment by dpgoetz** [comment]
looks good but a few typos

File: docs/rings.md (hunk 3 above)
Comment (corystone) [reply docs/rings.md]: need to be made, for example, adding or removing nodes or devices.

Comment (corystone) [reply docs/rings.md]: of the rings.

File: docs/rings.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
+# Ring Managment 101
+
+## Ring Overview
+
+The hummingbird rings are responsible for mapping what devices data is stored on.  There is a separate ring for the objects, containers and accounts. Rings are created when you build a new cluster and updated when changes to the cluster need to be made like for example, adding or removing nodes or devices to the storage cluster.  It is important that all servers in the cluster have the same and most current version of the ring.
+
+### Zones
+
+A Zone is an abstract concept in the ring that allows you to define failure domains within the cluster.  Each copy of an object will be stored in a device in a different zone.  It is recommended to choose the highest level of isolation that makes sense for the cluster size.  For example for a small cluster, with a handful of machines, each server might be a zone.  For a larger cluster, it might be each rack of servers, etc.
```
Comment (corystone) [inline docs/rings.md]: ", etc" probably shouldn't be there.

File: docs/rings.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+### Partitions
+
+When an object is stored in the cluster, it is not directly to a device, but instead to a partition via a hash.  The partitions are evenly distributed across all of the devices in the cluster.
```
Comment (corystone) [inline docs/rings.md]: it is not directly assigned to/placed on(?) a device. Or is "stored" the verb here?

File: docs/rings.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+The replica count determines how many copies of each object will be stored in the cluster.  A typical cluster will have 3 replicas, which means that the cluster will store 3 copies of the object on 3 different devices.
+
+### Partition Power
+
+The Partition Power is the number of partitions used by the ring.  A typical cluster will have a 4,194,304 (2^22) partitions and thus a partition power of 22.
+
+### Min Partition Hours
+
+When the state of the cluster changes, and data needs to be moved, it will try to only move 1 replica at a time (in a 3 replica cluster) to ensure that objects in the cluster remain available.  When a ring change is issued, the ring will not move partitions that have been moved in the Min Partition Hours setting.  This helps prevent the cluster from getting into a bad state through a series of ring changes.  A typical cluster will set the Min Partition Hours to 168 (1 week).
```
Comment (corystone) [inline docs/rings.md]: moved more recently than the Min Partition Hours

File: docs/rings.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+A ring can be generated with the command `hummingbird ring <builder_file> create <part_power> <replicas> <min_part_hours>` and an example object ring with the typical settings can be created with `hummingbird ring object.builder create 22 3 168`.
+
+The builder file contains extra information needed for building the rings and care should be taken to backup and not lose these files.
+
+## Adding Devices
+
+Each individual storage devices is stored in the ring.  Before adding the device, you will need to know the zone, server ip address and port, device name, and weight of the device.  The server ip address and port should be address and port that the server for that ring is running on.  The device name should be the name of the partitioned device.  The weight should be the number of Gigabytes the device can store.
```
Comment (corystone) [inline docs/rings.md]: individual storage device

Comment (corystone) [inline docs/rings.md]: should be the address and port

File: docs/rings.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+hummingbird ring account.builder add r1z3-10.1.1.12:6002/sdb1 500
+hummingbird ring account.builder add r1z3-10.1.1.12:6002/sdc1 500
+hummingbird ring account.builder rebalance
+```
+
+It is *very* important that you add all nodes and devices to the ring before running the rebalance command.
+
+## Handling a Device failure
+
+If a device has failed and will be replaced in the near future, then no ring changes are neccesary.  The system will work around the failure and replication will copy the data over when the drive is replaced.  Note that there will be some dedigration in performance until the device is replaced and fully replicated.
```
Comment (corystone) [inline docs/rings.md]: degradation

File: docs/rings.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+If a devices has failed and it will either not be replaced or will be a while before being replaced, then it would be a good idea to set the device's weight to 0.  The ring will spread the partitions that the device was responsible for to other devices in the cluster.  For example if device sde1 failed on node 10.1.1.11, the following commands would be issued:
```
Comment (corystone) [inline docs/rings.md]: device has failed

**Comment by creiht** [comment]
Thanks, they should all be updated now

**Pull request merged**